### DISCUSSION TITLE
Signal handling for container jobs

### DISCUSSION THREAD

**Discussion by jlordiales** [community/community#24932] [Actions]
<p>I’m sure I’m doing something wrong here but it seems like running a step inside a job container doesn’t really propagate signals to the process when the workflow is cancelled.</p>
<p>E.g., I have this straightforward go program that just runs until it receives a SIGINT or SIGTERM printing a message when it does: <a href="https://github.com/jlordiales/gha-signal-test/blob/267575885a59610973d0547679ddd4168af6b36c/main.go" class="inline-onebox" rel="noopener nofollow ugc">gha-signal-test/main.go at 267575885a59610973d0547679ddd4168af6b36c · jlordiales/gha-signal-test · GitHub</a></p>
<p>Now, if I run that on a workflow with <code>uses: docker://jlordiales/gha-signal-test:v1</code> and cancel it, it works as expected: <a href="https://github.com/jlordiales/gha-signal-test/runs/2079554004?check_suite_focus=true#step:3:6" class="inline-onebox" rel="noopener nofollow ugc">Add container job · jlordiales/gha-signal-test@2675758 · GitHub</a></p>
<p>If, however, I configure the workflow to use the exact same container as a job container:</p>
<pre><code class="lang-auto">jobs:
  test:
    name: Signal test
    runs-on: ubuntu-latest
    container: jlordiales/gha-signal-test:v1
    steps:
    - name: Run
      run: /bin/main
</code></pre>
<p>The process never receives the signal: <a href="https://github.com/jlordiales/gha-signal-test/runs/2079554001?check_suite_focus=true#step:3:7" class="inline-onebox" rel="noopener nofollow ugc">Add container job · jlordiales/gha-signal-test@2675758 · GitHub</a></p>
<p>Is there anything special I need to do so that this works as expected? I tried using <code>exec</code> in the <code>run</code> line so that my process replaces <code>bash</code> but still no dice. Also tried adding <code>--init</code> as a container option to ensure that that process runs with PID 1 and forwards all signals… still nothing.</p>
<p>What am I missing?</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>Stopping a container usually sends SIGTERM (and later SIGKILL) only to the main process of your container (and it looks like others just get killed if the whole container is). In case of the job container that’s <code>tail</code>, as you can see in the “Starting job container” part of the “Initialize containers” step of your job (note the <code>--entrypoint</code> option). I assume the idea is to have a process that does nothing except keeping the container open as an environment for the steps.</p>
<p><code>tail</code> is not know for forwarding signals to other processes, so if you want your process to receive a signal you’ll have to send it to the right process yourself, instead of relying on job cancellation.</p>

**Comment by jlordiales** [comment]
<p>good catch on that <code>tail -f</code> entrypoint… that really sucks. Apparently is not possible to override that with the <code>--entrypoint</code> flag in the <code>options</code> field either.</p>
<p>The only reason I wanted to use a job container instead of a proper action is because GitHub <em>still</em> doesn’t support docker actions from private registries (but they do for container jobs for some reason…).</p>
<p>Thanks for the answer!</p>

**Comment by Yikun** [comment]
I think I also hit this issue, from the history replies, it seems that there is no good/easy way to support cancel container jobs, right?
